servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -255,48 +335,68 @@ public void onComplete() {
                 // This is the case when SD has emitted some items but none of the hosts are active.
-                failed(StacklessNoAvailableHostException.newInstance(
-                    "No hosts are available to connect.", RoundRobinLoadBalancer.class, "selectConnection0(...)"));
+                failed(StacklessNoAvailableHostException.newInstance("No hosts are available to connect.",
+                    RoundRobinLoadBalancer.class, "selectConnection0(...)"));
         }
 
+        // try one loop over hosts and if all are expired, give up
         final int cursor = (indexUpdater.getAndIncrement(this) & Integer.MAX_VALUE) % activeHosts.size();
-        final Host<ResolvedAddress, C> host = activeHosts.get(cursor);
-        assert host != null : "Host can't be null.";
-        final ThreadLocalRandom rnd = ThreadLocalRandom.current();
-
-        // Try first to see if an existing connection can be used
-        final Object[] connections = host.connections;
-        // With small enough search space, attempt all connections.
-        // Back off after exploring most of the search space, it gives diminishing returns.
-        final int attempts = connections.length < MIN_SEARCH_SPACE ?
-                connections.length : (int) (connections.length * SEARCH_FACTOR);
-        for (int i = 0; i < attempts; i++) {
-            @SuppressWarnings("unchecked")
-            final C connection = (C) connections[rnd.nextInt(connections.length)];
-            if (selector.test(connection)) {
-                return succeeded(connection);
+        Host<ResolvedAddress, C> pickedHost = null;
+        for (int i = 0; i < activeHosts.size(); ++i) {
+            // for a particular iteration we maintain a local cursor without contention with other requests
+            int localCursor = (cursor + i) % activeHosts.size();
+            final Host<ResolvedAddress, C> host = activeHosts.get(localCursor);
+            assert host != null : "Host can't be null.";
+
+            final ThreadLocalRandom rnd = ThreadLocalRandom.current();
+
+            // Try first to see if an existing connection can be used
+            final Object[] connections = host.connections;
+            // With small enough search space, attempt all connections.
+            // Back off after exploring most of the search space, it gives diminishing returns.
+            final int attempts = connections.length < MIN_SEARCH_SPACE ?
+                    connections.length : (int) (connections.length * SEARCH_FACTOR);
+            for (int j = 0; j < attempts; ++j) {
+                @SuppressWarnings("unchecked")
+                final C connection = (C) connections[rnd.nextInt(connections.length)];
+                if (selector.test(connection)) {
+                    return succeeded(connection);
+                }
             }
-        }
 
+            // don't open new connections for expired hosts, try a different one
+            if (host.isActive()) {
```

Comment:
the name of the collection which holds all the SD events is called `activeHosts`, it is strange for this method to return false and still be retained in the collection, consider clarifying the naming to disambiguate (previously when hosts went inactive they were removed).

Reply:
I consolidated the naming - the collection was renamed `activeHosts` -> `usedHosts`, while the `AVAILABLE` state is represented consistently. Other states also have more consistent presence.